modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/exec/LogicalRelImplementor.java

```diff
@@ -406,6 +411,54 @@ public LogicalRelImplementor(
         }
     }
 
+    /** {@inheritDoc} */
+    @Override public Node<Row> visit(IgniteIndexCount rel) {
+        IgniteTable tbl = rel.getTable().unwrap(IgniteTable.class);
+        IgniteIndex idx = tbl.getIndex(rel.indexName());
+
+        if (idx != null && !tbl.isIndexRebuildInProgress()) {
+            Iterable<Row> src = new Iterable<Row>() {
+                @NotNull @Override public Iterator<Row> iterator() {
+                    RowFactory<Row> rowFactory = ctx.rowHandler().factory(ctx.getTypeFactory(), rel.getRowType());
+
+                    return Collections.singletonList(rowFactory.create(idx.unwrap(CacheIndexImpl.class)
+                        .scanCount(ctx))).iterator();
+                }
+            };
+
+            return new ScanNode<>(ctx, rel.getRowType(), src);
+        }
+        else {
+            CollectNode.Collector<Row> collector = new CollectNode.Collector<Row>(
+                ctx.rowHandler(),
+                ctx.rowHandler().factory(ctx.getTypeFactory(), rel.getRowType()), 1
+            ) {
+                /** */
+                private long cnt = 0;
+
+                @Override public void push(Row row) {
+                    ++cnt;
+                }
+
+                @Override public void clear() {
+                    cnt = 0;
+                }
+
+                @Override protected Object outData() {
+                    return cnt;
+                }
+            };
```

Comment:
Perhaps it's better to move this class inside the CollectNode class, WDYT?

Reply:
Fixed